File: sdks/python/kakeya/session.py
"""Kakeya Python SDK — :class:`Session` (PR-B4 of ADR 0008 Phase B).

A handle to one server-side session. Mirrors the ADR 0008 §2.2 RPC
surface as Python methods:

    session.append(token_ids)              -> AppendTokens
    session.generate(max_tokens=N)         -> Generate (server stream)
    session.info()                         -> GetSessionInfo
    session.close()                        -> CloseSession

After ``close()``, every method except ``session_id`` /
``info()`` raises :class:`~kakeya.errors.SessionClosedError`. The
runtime-side state may have been freed earlier (LRU / TTL eviction);
the SDK doesn't track that — the next RPC will raise
:class:`~kakeya.errors.SessionNotFoundError` if the runtime no
longer knows the id.
"""

from __future__ import annotations

from typing import Iterable, Iterator, Optional, TYPE_CHECKING

import grpc

from inference_engine.server.proto_gen.kakeya.v1 import (
    runtime_pb2,
    runtime_pb2_grpc,
)
from kakeya.errors import (
    SessionClosedError,
    _wrap_grpc_error,
)

if TYPE_CHECKING:
    from kakeya.client import Client


class SessionInfo:
    """Read-only snapshot of a session's server-side state.

    Returned by :meth:`Session.info`. Field names mirror
    :class:`runtime_pb2.GetSessionInfoResponse` for direct
    correspondence with the wire contract.
    """

    __slots__ = (
        "history_length",
        "kv_live_bytes",
        "cache_invariant_inv1_violations",
        "cache_invariant_inv2_violations",
        "idle_seconds",
    )

    def __init__(
        self,
        *,
        history_length: int,
        kv_live_bytes: int,
        cache_invariant_inv1_violations: int,
        cache_invariant_inv2_violations: int,
        idle_seconds: float,
    ) -> None:
        self.history_length = history_length
        self.kv_live_bytes = kv_live_bytes
        self.cache_invariant_inv1_violations = cache_invariant_inv1_violations
        self.cache_invariant_inv2_violations = cache_invariant_inv2_violations
        self.idle_seconds = idle_seconds

    def __repr__(self) -> str:
        return (
            f"SessionInfo(history_length={self.history_length}, "
            f"kv_live_bytes={self.kv_live_bytes}, "
            f"inv1={self.cache_invariant_inv1_violations}, "
            f"inv2={self.cache_invariant_inv2_violations}, "
            f"idle_seconds={self.idle_seconds:.3f})"
        )


class Session:
    """One server-side session, addressable by ``session_id``."""

    def __init__(self, *, client: "Client", session_id: str) -> None:
        self._client = client
        # Prefer the client's stub (already constructed) over making
        # a new one; this keeps a single channel per Client instance.
        self._stub: runtime_pb2_grpc.RuntimeServiceStub = client._stub
        self._session_id = session_id
        self._closed = False
        # Populated after the most recent generate() call returns.
        # Useful for callers that want the stop_reason / token count
        # without having to wrap iteration in their own bookkeeping.
        self._last_stop_reason: Optional[int] = None
        self._last_generated_token_count: int = 0
        self._last_prefill_duration_seconds: float = 0.0
        self._last_total_duration_seconds: float = 0.0
        self._last_history_truncated_dropped: Optional[int] = None

    # ------------------------------------------------------------------
    # Properties
    # ------------------------------------------------------------------

    @property
    def session_id(self) -> str:
        return self._session_id

    @property
    def closed(self) -> bool:
        return self._closed

    @property
    def last_stop_reason(self) -> Optional[int]:
        """``runtime_pb2.GenerateDone.StopReason`` enum value from the
        most recent :meth:`generate` call, or ``None`` if no call has
        been made yet."""
        return self._last_stop_reason

    @property
    def last_generated_token_count(self) -> int:
        return self._last_generated_token_count

    @property
    def last_prefill_duration_seconds(self) -> float:
        return self._last_prefill_duration_seconds

    @property
    def last_total_duration_seconds(self) -> float:
        return self._last_total_duration_seconds

    @property
    def last_history_truncated_dropped(self) -> Optional[int]:
        """If the most recent :meth:`generate` started in
        sink+window-truncated mode, the runtime emitted a
        ``HistoryTruncated`` event with the number of dropped
        tokens — this property exposes that count. ``None`` if the
        last call did not encounter truncation, or if no call has
        been made yet."""
        return self._last_history_truncated_dropped

    # ------------------------------------------------------------------
    # RPC methods
    # ------------------------------------------------------------------

    def append(self, token_ids: Iterable[int]) -> int:
        """Append raw token ids to the session's history.

        Returns the new ``history_length``. Empty input is a
        runtime-side no-op.
        """
        self._check_open()
        request = runtime_pb2.AppendTokensRequest(
            session_id=self._session_id,
            token_ids=list(token_ids),
        )
        try:
            response = self._stub.AppendTokens(request)
        except grpc.RpcError as exc:
            raise _wrap_grpc_error(exc) from exc
        return response.history_length

    def generate(
        self,
        *,
        max_tokens: int,
        seed: Optional[int] = None,
        temperature: Optional[float] = None,
        top_p: Optional[float] = None,
        top_k: Optional[int] = None,
    ) -> Iterator[int]:
        """Stream generated token ids.

        Yields ``int`` token ids in generation order. The iterator
        is exhausted when the server emits a ``GenerateDone``
        frame; metadata (stop reason, count, durations) is then
        available via :attr:`last_stop_reason` etc.

        v0.3 supports only greedy decoding. Setting
        ``temperature`` / ``top_p`` / ``top_k`` to anything other
        than the greedy no-op default raises
        :class:`~kakeya.errors.InvalidArgumentError` from the
        runtime (per ADR 0008 §2.10 "no graceful degradation").
        """
        self._check_open()
        # Reset metadata from any prior call before this one starts.
        self._last_stop_reason = None
        self._last_generated_token_count = 0
        self._last_prefill_duration_seconds = 0.0
        self._last_total_duration_seconds = 0.0
        self._last_history_truncated_dropped = None

        request = runtime_pb2.GenerateRequest(
            session_id=self._session_id,
            max_tokens=max_tokens,
        )
        if seed is not None:
            request.seed = seed
        if temperature is not None:
            request.temperature = temperature
        if top_p is not None:
            request.top_p = top_p
        if top_k is not None:
            request.top_k = top_k

        try:
            for response in self._stub.Generate(request):
                payload = response.WhichOneof("payload")
                if payload == "token_id":
                    yield response.token_id
                elif payload == "truncated":
                    self._last_history_truncated_dropped = (
                        response.truncated.dropped_token_count
                    )
                elif payload == "done":
                    done = response.done
                    self._last_stop_reason = done.stop_reason
                    self._last_generated_token_count = (
                        done.generated_token_count
                    )
                    self._last_prefill_duration_seconds = (
                        done.prefill_duration_seconds
                    )
                    self._last_total_duration_seconds = (
                        done.total_duration_seconds
                    )
                    return
        except grpc.RpcError as exc:
            raise _wrap_grpc_error(exc) from exc

    def info(self) -> SessionInfo:
        """Return a snapshot of the session's server-side state.

        Allowed even after :meth:`close` has been called locally —
        in that case the call goes to the runtime and most likely
        returns :class:`~kakeya.errors.SessionNotFoundError`.
        """
        request = runtime_pb2.GetSessionInfoRequest(
            session_id=self._session_id,
        )
        try:
            response = self._stub.GetSessionInfo(request)
        except grpc.RpcError as exc:
            raise _wrap_grpc_error(exc) from exc
        return SessionInfo(
            history_length=response.history_length,
            kv_live_bytes=response.kv_live_bytes,
            cache_invariant_inv1_violations=(
                response.cache_invariant_inv1_violations
            ),
            cache_invariant_inv2_violations=(
                response.cache_invariant_inv2_violations
            ),
            idle_seconds=response.idle_seconds,
        )

    def close(self) -> int:
        """Close the session on the runtime. Returns the final
        ``history_length``.

        Idempotent at the SDK level: a second call returns 0
        without contacting the runtime, the same way ``Client.close()``
        is idempotent. (A first close that fails on the wire still
        flips the local closed flag — the SDK assumes the runtime
        is unreachable rather than dribbling out further calls.)
        """
        if self._closed:
            return 0
        request = runtime_pb2.CloseSessionRequest(
            session_id=self._session_id,
        )
        try:
            response = self._stub.CloseSession(request)
        except grpc.RpcError as exc:
            self._closed = True
            raise _wrap_grpc_error(exc) from exc
        self._closed = True
        return response.final_history_length

    # ------------------------------------------------------------------
    # Context manager
    # ------------------------------------------------------------------

    def __enter__(self) -> "Session":
        return self

    def __exit__(self, exc_type, exc, tb) -> None:
        # Best-effort close on context exit; swallow SessionNotFoundError
        # because the runtime may have evicted the session between the
        # last RPC and this exit.
        try:
            self.close()
        except Exception:  # pragma: no cover - best-effort cleanup
            pass

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    def _check_open(self) -> None:
        if self._closed:
            raise SessionClosedError(
                f"session {self._session_id!r} has been closed locally; "
                "create a new session to continue work",
            )

File: sdks/typescript/src/session.ts
/**
 * Kakeya TypeScript SDK — Session (PR-B5 of ADR 0008 Phase B).
 *
 * Mirrors sdks/python/kakeya/session.py symbol-for-symbol.
 *
 * @public
 */

import * as grpc from "@grpc/grpc-js";

import {
  SessionClosedError,
  wrapGrpcError,
} from "./errors.js";
import {
  AppendTokensRequest,
  CloseSessionRequest,
  GenerateDone_StopReason,
  GenerateRequest,
  GenerateResponse,
  GetSessionInfoRequest,
} from "./proto_gen/kakeya/v1/runtime.js";
import type { Client } from "./client.js";

/**
 * Read-only snapshot of a session's server-side state. Returned by
 * {@link Session.info}.
 */
export interface SessionInfo {
  /** Number of tokens currently in the session's history. */
  historyLength: number;
  /** Live KV bytes held by this session's slab. */
  kvLiveBytes: number;
  /** INV-1 violations observed (must be 0 in healthy operation). */
  cacheInvariantInv1Violations: number;
  /** INV-2 violations observed (must be 0 in healthy operation). */
  cacheInvariantInv2Violations: number;
  /** Wall seconds since the session's last RPC interaction. */
  idleSeconds: number;
}

/** Options accepted by {@link Session.generate}. */
export interface GenerateOptions {
  /** Maximum tokens to emit in this Generate call. Required, must be >= 1. */
  maxTokens: number;
  /** Per-Generate sampling seed (greedy mode ignores it; accepted for OQ-4). */
  seed?: number | string;
  /** Sampling temperature. v0.3 supports only 0 (greedy). */
  temperature?: number;
  /** Top-p sampling. v0.3 supports only unset (greedy). */
  topP?: number;
  /** Top-k sampling. v0.3 supports only 1 (greedy). */
  topK?: number;
}

/** Metadata populated after a {@link Session.generate} stream finishes. */
export interface GenerateResult {
  stopReason: GenerateDone_StopReason | undefined;
  generatedTokenCount: number;
  prefillDurationSeconds: number;
  totalDurationSeconds: number;
  /**
   * Number of tokens dropped by the runtime's sink+window cache at
   * the start of this Generate call, or `undefined` if no
   * truncation occurred. Per the proto contract, the runtime emits
   * the truncated frame at most once per call, before any
   * `tokenId`.
   */
  historyTruncatedDropped: number | undefined;
}

/** A handle to one server-side session. */
export class Session {
  readonly sessionId: string;
  private readonly _client: Client;
  private _closed: boolean = false;
  private _lastResult: GenerateResult = {
    stopReason: undefined,
    generatedTokenCount: 0,
    prefillDurationSeconds: 0,
    totalDurationSeconds: 0,
    historyTruncatedDropped: undefined,
  };

  /** @internal — only :class:`Client.createSession` should call this. */
  constructor(client: Client, sessionId: string) {
    this._client = client;
    this.sessionId = sessionId;
  }

  /** True once {@link close} has been called locally. */
  get closed(): boolean {
    return this._closed;
  }

  /**
   * Result metadata from the most recent {@link generate} call.
   * Populated as the stream completes; reset at the start of the
   * next call.
   */
  get lastResult(): GenerateResult {
    return this._lastResult;
  }

  /**
   * Append raw token ids to the session's history. Returns the new
   * `historyLength`.
   */
  async append(tokenIds: number[]): Promise<number> {
    this._checkOpen();
    const request: AppendTokensRequest = {
      sessionId: this.sessionId,
      tokenIds,
    };
    return new Promise<number>((resolve, reject) => {
      this._client.grpcClient.appendTokens(request, (err, response) => {
        if (err) {
          reject(wrapGrpcError(err));
          return;
        }
        resolve(Number(response.historyLength));
      });
    });
  }

  /**
   * Stream generated token ids as an async iterable. Each iterator
   * yield is one committed token. After iteration completes,
   * {@link lastResult} is populated with the {@link GenerateResult}
   * metadata (stop reason, count, durations, truncation).
   *
   * v0.3 supports only greedy decoding. Setting `temperature` /
   * `topP` / `topK` to anything other than the greedy no-op default
   * raises {@link InvalidArgumentError} from the runtime.
   */
  generate(opts: GenerateOptions): AsyncIterable<number> {
    this._checkOpen();
    // Reset metadata at the start of every call so stale values
    // from a previous call don't leak into the next.
    this._lastResult = {
      stopReason: undefined,
      generatedTokenCount: 0,
      prefillDurationSeconds: 0,
      totalDurationSeconds: 0,
      historyTruncatedDropped: undefined,
    };
    // ts-proto generates `seed: string | undefined` because the
    // proto field is `optional uint64` and we set forceLong=string
    // (JS `number` can't safely represent the full uint64 range).
    // Accept number for ergonomics and stringify; callers needing
    // the full range pass a string directly.
    const request: GenerateRequest = {
      sessionId: this.sessionId,
      maxTokens: opts.maxTokens,
      seed: opts.seed === undefined ? undefined : String(opts.seed),
      temperature: opts.temperature,
      topP: opts.topP,
      topK: opts.topK,
    };
    const stream = this._client.grpcClient.generate(request);
    return this._consumeGenerateStream(stream);
  }

  private async *_consumeGenerateStream(
    stream: grpc.ClientReadableStream<GenerateResponse>,
  ): AsyncIterable<number> {
    // Bridge the event-emitter stream into an async iterator. We
    // buffer (data | error | end) events and replay them in order
    // so the caller can `for await` cleanly. Errors map to typed
    // KakeyaError subclasses before being thrown.
    type Event =
      | { kind: "data"; response: GenerateResponse }
      | { kind: "error"; err: grpc.ServiceError }
      | { kind: "end" };

    const events: Event[] = [];
    let resolveNext: (() => void) | null = null;
    const wakeup = () => {
      if (resolveNext) {
        const r = resolveNext;
        resolveNext = null;
        r();
      }
    };

    stream.on("data", (response: GenerateResponse) => {
      events.push({ kind: "data", response });
      wakeup();
    });
    stream.on("error", (err: grpc.ServiceError) => {
      events.push({ kind: "error", err });
      wakeup();
    });
    stream.on("end", () => {
      events.push({ kind: "end" });
      wakeup();
    });

    while (true) {
      while (events.length > 0) {
        const ev = events.shift()!;
        if (ev.kind === "error") {
          throw wrapGrpcError(ev.err);
        }
        if (ev.kind === "end") {
          return;
        }
        const r = ev.response;
        if (r.tokenId !== undefined) {
          yield r.tokenId;
        } else if (r.truncated !== undefined) {
          this._lastResult.historyTruncatedDropped = Number(
            r.truncated.droppedTokenCount,
          );
        } else if (r.done !== undefined) {
          this._lastResult.stopReason = r.done.stopReason;
          this._lastResult.generatedTokenCount = r.done.generatedTokenCount;
          this._lastResult.prefillDurationSeconds =
            r.done.prefillDurationSeconds;
          this._lastResult.totalDurationSeconds = r.done.totalDurationSeconds;
        }
      }
      // Wait for the next event.
      await new Promise<void>((resolve) => {
        resolveNext = resolve;
      });
    }
  }

  /** Return a snapshot of the session's server-side state. */
  async info(): Promise<SessionInfo> {
    const request: GetSessionInfoRequest = { sessionId: this.sessionId };
    return new Promise<SessionInfo>((resolve, reject) => {
      this._client.grpcClient.getSessionInfo(request, (err, response) => {
        if (err) {
          reject(wrapGrpcError(err));
          return;
        }
        resolve({
          historyLength: Number(response.historyLength),
          kvLiveBytes: Number(response.kvLiveBytes),
          cacheInvariantInv1Violations: Number(
            response.cacheInvariantInv1Violations,
          ),
          cacheInvariantInv2Violations: Number(
            response.cacheInvariantInv2Violations,
          ),
          idleSeconds: response.idleSeconds,
        });
      });
    });
  }

  /**
   * Close the session on the runtime. Returns the final history
   * length. Idempotent at the SDK level: a second call returns 0
   * without contacting the runtime.
   */
  async close(): Promise<number> {
    if (this._closed) {
      return 0;
    }
    const request: CloseSessionRequest = { sessionId: this.sessionId };
    return new Promise<number>((resolve, reject) => {
      this._client.grpcClient.closeSession(request, (err, response) => {
        if (err) {
          this._closed = true;
          reject(wrapGrpcError(err));
          return;
        }
        this._closed = true;
        resolve(Number(response.finalHistoryLength));
      });
    });
  }

  private _checkOpen(): void {
    if (this._closed) {
      throw new SessionClosedError(
        `session '${this.sessionId}' has been closed locally; ` +
          "create a new session to continue work",
      );
    }
  }
}
